File: backend/src/main/java/com/taskmanager/task/Quadrant.java
package com.taskmanager.task;

public enum Quadrant {
    DO_FIRST(Importance.IMPORTANT, Urgency.URGENT),
    SCHEDULE(Importance.IMPORTANT, Urgency.NOT_URGENT),
    DELEGATE(Importance.NOT_IMPORTANT, Urgency.URGENT),
    DROP(Importance.NOT_IMPORTANT, Urgency.NOT_URGENT);

    private final Importance importance;
    private final Urgency urgency;

    Quadrant(Importance importance, Urgency urgency) {
        this.importance = importance;
        this.urgency = urgency;
    }

    public Importance importance() {
        return importance;
    }

    public Urgency urgency() {
        return urgency;
    }

    public static Quadrant of(Importance importance, Urgency urgency) {
        for (Quadrant quadrant : values()) {
            if (quadrant.importance == importance && quadrant.urgency == urgency) {
                return quadrant;
            }
        }
        throw new IllegalArgumentException("no quadrant for " + importance + " and " + urgency);
    }
}

File: frontend/src/features/tasks/quadrant.ts
import type { Importance, Quadrant, Urgency } from '../../types/task'

export interface Axes {
  importance: Importance
  urgency: Urgency
}

export interface QuadrantCell extends Axes {
  quadrant: Quadrant
}

/**
 * One table, both directions. A cell IS an {importance, urgency} pair, so a contradictory
 * combination is not representable from the UI — mirroring the backend, where two separate
 * enum columns make it impossible to store. The error is unreachable, not merely validated.
 *
 * Layout from the kit: urgency across columns, importance down rows.
 *
 *              URGENT      NOT_URGENT
 *   IMPORTANT  Do first    Schedule
 *   NOT_IMP.   Delegate    Drop
 */
export const QUADRANT_CELLS: QuadrantCell[] = [
  { quadrant: 'DO_FIRST', importance: 'IMPORTANT', urgency: 'URGENT' },
  { quadrant: 'SCHEDULE', importance: 'IMPORTANT', urgency: 'NOT_URGENT' },
  { quadrant: 'DELEGATE', importance: 'NOT_IMPORTANT', urgency: 'URGENT' },
  { quadrant: 'DROP', importance: 'NOT_IMPORTANT', urgency: 'NOT_URGENT' },
]

export function quadrantOf(importance: Importance, urgency: Urgency): Quadrant {
  const match = QUADRANT_CELLS.find(
    (cell) => cell.importance === importance && cell.urgency === urgency,
  )
  if (!match) {
    throw new Error(`no quadrant for ${importance} and ${urgency}`)
  }
  return match.quadrant
}

export function axesOf(quadrant: Quadrant): Axes {
  const match = QUADRANT_CELLS.find((cell) => cell.quadrant === quadrant)
  if (!match) {
    throw new Error(`unknown quadrant ${quadrant}`)
  }
  return { importance: match.importance, urgency: match.urgency }
}
